src/platform/storage/zephyr/nvs/storage.c

```diff
@@ -163,41 +211,85 @@ crc_check(const unsigned char *data, const size_t data_len) {
+        return MENDER_FAIL;
+    }
+#endif /* CONFIG_MENDER_STORAGE_PARTITION_CUSTOM */
 
     /* Get flash info */
-    mender_storage_nvs_handle.flash_device = MENDER_STORAGE_DEVICE;
+    /** First check if the device is ready. */
     if (!device_is_ready(mender_storage_nvs_handle.flash_device)) {
         mender_log_error("Flash device not ready");
         return MENDER_FAIL;
     }
 
-    n_sectors = sizeof(flash_sectors) / sizeof(flash_sectors[0]);
-    result    = flash_area_get_sectors(MENDER_STORAGE_FLASH_AREA, &n_sectors, flash_sectors);
+    /** Then get the information about the partition offset on the device
+        (flash_sectors[0].fs_off below is an offset relative to the partition
+        start). */
+    if (0 != (result = flash_area_open(part_id, &fap))) {
+        mender_log_error("Unable to open the Mender storage flash area");
+        return MENDER_FAIL;
+    }
+    part_offset = fap->fa_off;
+    flash_area_close(fap);
+
+    /** Then get information about the individual sectors. */
+    const uint32_t n_sectors_required = n_sectors; /* optimized out for sure */
+    result                            = flash_area_get_sectors(part_id, &n_sectors, flash_sectors);
     if ((0 != result) && (-ENOMEM != result)) {
         /* -ENOMEM means there were more sectors than the supplied size of the
             sector array (flash_sectors), but we don't worry about that, we just
             need the info about the first N sectors we want to use. */
         mender_log_error("Failed to get info about flash sectors in the Mender flash area [%d]", -result);
         return MENDER_FAIL;
     }
-    if (n_sectors != (sizeof(flash_sectors) / sizeof(flash_sectors[0]))) {
-        mender_log_error(
-            "Not enough sectors on the flash for Mender (required: %" PRIu32 "d, available: %zu)", sizeof(flash_sectors) / sizeof(flash_sectors[0]), n_sectors);
+    if (n_sectors != n_sectors_required) {
+        mender_log_error("Not enough sectors on the flash for Mender (required: %" PRIu32 "d, available: %" PRIu32 "d)", n_sectors_required, n_sectors);
         return MENDER_FAIL;
     }
 
-    mender_storage_nvs_handle.offset       = MENDER_STORAGE_OFFSET;
-    mender_storage_nvs_handle.sector_size  = (uint16_t)flash_sectors[0].fs_size;
+    /** Take sector size from the first sector. */
+    mender_storage_nvs_handle.offset       = part_offset + flash_sectors[sector_offset].fs_off;
+    mender_storage_nvs_handle.sector_size  = (uint16_t)flash_sectors[sector_offset].fs_size;
     mender_storage_nvs_handle.sector_count = CONFIG_MENDER_STORAGE_NVS_SECTOR_COUNT;
 
     /* Mount NVS */
     if (0 != (result = nvs_mount(&mender_storage_nvs_handle))) {
         mender_log_error("Unable to mount NVS storage, result = %d", result);
         return MENDER_FAIL;
     }
-    mender_log_debug("Initialized Mender NVS with %u sectors (%zu bytes available)",
+    mender_log_debug("Initialized Mender NVS at 0x%jx with %u sectors (%zu bytes available)",
```

Comment:
Have you verified that the length modifier `j` is available?

Reply:
Yes. And I think it's a pretty standard one, part of the `printf(3)` man page.